### PR TITLE
continuum_mechanics: Added plotting methods

### PR DESCRIPTION
**Pull request by jashan498**
#### Brief description of what is fixed or changed
Methods to plot Shear force, Bending moment, Slope and Delfection has been added. 
`plot_all` was also added which plots Shear force, Bending moment and Delfection on a single graph.

[Here](https://github.com/jashan498/beam_examples/blob/master/beam_plots.ipynb) is an example of how plots look.

#### Release Notes

* physics.continuum_mechanics
  * added methods to plot shear force, bending moment, slope and deflection.

### CODE REVIEW

**Comment by asmeurer** [comment]
I think this could use a release notes entry. Put it under `* physics.continuum_mechanics`.

**Comment by jashan498** [comment]
Why docsrting of methods containing `plot` failing? For example, one of the test failures is:
```
_________ sympy.physics.continuum_mechanics.beam.Beam.plot_shear_force _________
File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy/physics/continuum_mechanics/beam.py", line 1151, in sympy.physics.continuum_mechanics.beam.Beam.plot_shear_force
Failed example:
    b.plot_shear_force()
Expected:
    Plot object containing:
        [0]: cartesian line: -13750*SingularityFunction(x, 0, 0) + 5000*SingularityFunction(x, 2, 0)
        + 10000*SingularityFunction(x, 4, 1) - 31250*SingularityFunction(x, 8, 0)
        - 10000*SingularityFunction(x, 8, 1) for x over (0.0, 8.0)
Got:
    29795.5 |                                                      / 
            |                                                    ..  
            |                                                   /    
            |                                                 ..     
            |                                               ..       
            |                                              /         
            |                                            ..          
            |                                          ..            
    8022.73 | ----------------------------------------/--------------
            |                                       ..               
            |                                     ..                 
            |                                    /                   
            |                                  ..                    
            |                                ..                      
            |                               /                        
            |               ................                         
            |                                                        
     -13750 | ..............                                         
              0                      4                          8
    Plot object containing:
    [0]: cartesian line: -13750*SingularityFunction(x, 0, 0) + 5000*SingularityFunction(x, 2, 0) + 10000*SingularityFunction(x, 4, 1) - 31250*SingularityFunction(x, 8, 0) - 10000*SingularityFunction(x, 8, 1) for x over (0.0, 8.0)
```
But `[0]: cartesian line` seems identical for both `Expected` and `Got`. 
Also `bin/doctest beam` is passing all docstrings locally.

**Comment by jashan498** [comment]
Ping @parsoyaarihant @moorepants

File: sympy/physics/continuum_mechanics/beam.py [hunk 1]
```diff
@@ -1116,6 +1118,267 @@ def max_deflection(self):
         else:
             return None
 
+    @doctest_depends_on(modules=('matplotlib',))
```
Comment (jashan498) [inline sympy/physics/continuum_mechanics/beam.py]: There is no need to use `doctest_depend_on` after [#14984](https://github.com/sympy/sympy/pull/14984).

File: sympy/physics/continuum_mechanics/beam.py [hunk 2]
```diff
@@ -1116,6 +1117,262 @@ def max_deflection(self):
         else:
             return None
 
+    def plot_shear_force(self, var={}):
+        """
+        Returns a plot for Shear force present in the Beam object.
+
+        Parameters
+        ==========
+        var : dictionary
```
Comment (moorepants) [inline sympy/physics/continuum_mechanics/beam.py]: Maybe call this `subs` instead of `var` as it resembles the `subs()` method on SymPy expressions.

File: sympy/physics/continuum_mechanics/beam.py [hunk 3]
```diff
@@ -1116,6 +1117,262 @@ def max_deflection(self):
...
+            length = self.length
+        p = plot(shear_force.subs(var), bending_moment.subs(var), deflection.subs(var),
+                (self.variable, 0, length), show=False)
+        labels = ['Shear Force', 'Bending Moment', 'Deflection']
+        colors = ['g', 'b', 'r']
+        for n, line in enumerate(p):
+            line.label= labels[n]
+            line.line_color = colors[n]
+        p.legend = True
+        return p.show()
```
Comment (moorepants) [inline sympy/physics/continuum_mechanics/beam.py]: Always return the matplotlib axes and do not call show. It is convention to let the user call show.

File: sympy/physics/continuum_mechanics/beam.py [hunk 4]
```diff
@@ -1116,6 +1117,262 @@ def max_deflection(self):
...
+            if sym not in var:
+                raise ValueError('Value of %s was not passed.' %sym)
+        if self.length in var:
+            length = var[self.length]
+        else:
+            length = self.length
+        return plot(deflection.subs(var), (self.variable, 0, length), title='Deflection',
+                xlabel='position', ylabel='Value', line_color='r')
+
+    def plot_all(self, var={}):
```
Comment (moorepants) [inline sympy/physics/continuum_mechanics/beam.py]: I recommend that this create a subplot since the units are different for all of them.

**Review by moorepants — CHANGES_REQUESTED** [review]
This looks great! Very useful convenience methods. Please address the few comments i have.

File: sympy/physics/continuum_mechanics/beam.py (hunk 4 above)
Comment (moorepants) [reply sympy/physics/continuum_mechanics/beam.py]: Would simply `plot()` be a sufficient name?

Comment (jashan498) [reply sympy/physics/continuum_mechanics/beam.py]: One `plot` method already exists in `plotting` module. Is it okay to give same name to this method?

Comment (moorepants) [reply sympy/physics/continuum_mechanics/beam.py]: I don't see why not. 

But now that I think about this, here we have a "beam". If I were a new user and saw a "plot" method I guess I'd think it would show a diagram of the beam (like what you see in the problem description in engr. books). If I saw "plot_all" I'd probably guess it'd plot multiple things but it isn't that informative about what will be plotted. Maybe "plot" should be reserved for showing a diagram of the beam and "plot_diagrams" or "plot_all" or something better would be for this method. This method you have plots a "shear diagram", "moment diagram", "slope curve", and "deflection curve". Maybe "plot_loaded_results" or "plot_loading_results"...not thinking of a great concise name.

**Comment by jashan498** [comment]
@moorepants I used `matplotlib` and `numpy` straightaway in `plot_loading_results` method rather than using SymPy's `plot` as `plot` was unable to make subplots (I even tried using `_backend`).

This is how resultant plot looks like now:
```
>>> from sympy.physics.continuum_mechanics.beam import Beam
>>> from sympy import symbols
>>> R1, R2 = symbols('R1, R2')
>>> b = Beam(8, 200*(10**9), 400*(10**-6))
>>> b.apply_load(5000, 2, -1)
>>> b.apply_load(R1, 0, -1)
>>> b.apply_load(R2, 8, -1)
>>> b.apply_load(10000, 4, 0, end=8)
>>> b.bc_deflection = [(0, 0), (8, 0)]
>>> b.solve_for_reaction_loads(R1, R2)
>>> plt = b.plot_loading_results()
>>> plt.show()
```
loading_results

**Comment by moorepants** [comment]
The subplot looks good. Do you think a 4 x1 plot would be better than a 2x2? All of the plots share the x axis (beam length) and in text books you often see them plotted 4x1.

File: sympy/physics/continuum_mechanics/beam.py [hunk 5]
```diff
@@ -1116,9 +1118,289 @@ def max_deflection(self):
         else:
             return None
 
+    def plot_shear_force(self, subs={}):
```
Comment (moorepants) [inline sympy/physics/continuum_mechanics/beam.py]: I recommend `subs=None`. Having mutables as a default arg can sometimes cause confusion.

File: sympy/physics/continuum_mechanics/beam.py [hunk 6]
```diff
@@ -1116,9 +1118,289 @@ def max_deflection(self):
...
+        >>> from sympy import symbols
+        >>> R1, R2 = symbols('R1, R2')
+        >>> b = Beam(8, 200*(10**9), 400*(10**-6))
+        >>> b.apply_load(5000, 2, -1)
+        >>> b.apply_load(R1, 0, -1)
+        >>> b.apply_load(R2, 8, -1)
+        >>> b.apply_load(10000, 4, 0, end=8)
+        >>> b.bc_deflection = [(0, 0), (8, 0)]
+        >>> b.solve_for_reaction_loads(R1, R2)
+        >>> b.plot_shear_force()
```
Comment (moorepants) [inline sympy/physics/continuum_mechanics/beam.py]: What happens if you call the plot methods and the beam has not been solved yet? i.e. you've only done:

```python
b = Beam()
b.plot_shear_force()
```

File: sympy/physics/continuum_mechanics/beam.py [hunk 7]
```diff
@@ -1116,9 +1118,289 @@ def max_deflection(self):
...
+        >>> b.apply_load(R2, 8, -1)
+        >>> b.apply_load(10000, 4, 0, end=8)
+        >>> b.bc_deflection = [(0, 0), (8, 0)]
+        >>> b.solve_for_reaction_loads(R1, R2)
+        >>> b.plot_loading_results()
+        Figure(640x480)
+        """
+        import matplotlib.pyplot as plt
+        from numpy import linspace
+        from sympy import lambdify
```
Comment (moorepants) [inline sympy/physics/continuum_mechanics/beam.py]: Import these at the top with `sympy.external.import_module()`.

File: sympy/physics/continuum_mechanics/beam.py [hunk 8]
```diff
@@ -1116,9 +1118,289 @@ def max_deflection(self):
...
+        axs[0, 0].set_title("Shear Force")
+        axs[0, 1].plot(points, moment(points), color='b')
+        axs[0, 1].set_title("Bending Moment")
+        axs[1, 0].plot(points, slope(points), color='m')
+        axs[1, 0].set_title("Slope")
+        axs[1, 1].plot(points, deflection(points), color='r')
+        axs[1, 1].set_title("Deflection")
+        fig.tight_layout()    # For better spacing between subplots
+
+        return fig
```
Comment (moorepants) [inline sympy/physics/continuum_mechanics/beam.py]: It is convention to return `axs` not `fig`.

File: sympy/physics/continuum_mechanics/beam.py [hunk 9]
```diff
@@ -1138,10 +1420,10 @@ class Beam_3d(Beam):
     Beam is fixed at both of its end. So, deflection of the beam at the both ends
     is restricted.
 
-    >>> from sympy.physics.continuum_mechanics.beam import Beam_3d
+    >>> from sympy.physics.continuum_mechanics.beam import Beam3D
     >>> from sympy import symbols
     >>> l, E, G, I, A = symbols('l, E, G, I, A')
-    >>> b = Beam_3d(l, E, G, I, A)
+    >>> b = Beam3D(l, E, G, I, A)
```
Comment (moorepants) [inline sympy/physics/continuum_mechanics/beam.py]: It would be better to make this change in a separate PR.

File: sympy/physics/continuum_mechanics/beam.py (hunk 6 above)
Comment (jashan498) [reply sympy/physics/continuum_mechanics/beam.py]: It returns only co-ordinate axes without a plot as there is no load present on the beam.

File: sympy/physics/continuum_mechanics/beam.py (hunk 8 above)
Comment (jashan498) [reply sympy/physics/continuum_mechanics/beam.py]: but `type(axs) = <class 'numpy.ndarray'>`, we can't use `.show()` on it to see the plot?

**Comment by jashan498** [comment]
This is how 4x1 looks like now:
load

**Comment by moorepants** [comment]
I like the 4x1 better, how bout you? I makes the beams seem longer and slender, which is what they are.

**Comment by moorepants** [comment]
One more note. The deflection curves are traditionally show with forces applied pointing down and the be deflecting downwards. I'm sure this is a sign convention thing.

**Comment by jashan498** [comment]
Initially, I thought `4x1` is too long horizontally but yeah you are right they represent beam's length so they should be like that. Also its easier to read than `2x2`.

Yes, the deflection curve is upwards in this case because while applying load I treated downward direction positive (See its downward in [this](https://github.com/jashan498/beam_examples/blob/master/beam_plots.ipynb) example though).

**Comment by moorepants** [comment]
Here is a typical figure from a textbook:

selection_218

I suggest that you work to emulate this wrt to sign conventions, etc.

**Comment by jashan498** [comment]
Have a look at [this](https://github.com/jashan498/beam_examples/blob/master/beam_plot_2.ipynb) notebook.

**Comment by jashan498** [comment]
Btw `plot_resulting_loads` is creating two plots in jupyter notebook everytime. Creates only one when used in terminal.

**Comment by moorepants** [comment]
Don't return the figure, return the axes in that function.

File: sympy/physics/continuum_mechanics/beam.py [hunk 10]
```diff
@@ -1116,6 +1119,298 @@ def max_deflection(self):
...
+        axs[0].set_title("Shear Force")
+        axs[1].plot(points, moment(points), color='b')
+        axs[1].set_title("Bending Moment")
+        axs[2].plot(points, slope(points), color='m')
+        axs[2].set_title("Slope")
+        axs[3].plot(points, deflection(points), color='r')
+        axs[3].set_title("Deflection")
+        fig.tight_layout()    # For better spacing between subplots
+
+        return fig
```
Comment (moorepants) [inline sympy/physics/continuum_mechanics/beam.py]: Change to `return axs`.

File: sympy/physics/continuum_mechanics/beam.py [hunk 11]
```diff
@@ -1116,6 +1119,298 @@ def max_deflection(self):
...
+        # Creating a grid for subplots with 2 rows and 2 columns
+        fig, axs = plt.subplots(4, 1)
+        # axs is a 2D-numpy array containing axes
+        axs[0].plot(points, shear(points), color='g')
+        axs[0].set_title("Shear Force")
+        axs[1].plot(points, moment(points), color='b')
+        axs[1].set_title("Bending Moment")
+        axs[2].plot(points, slope(points), color='m')
+        axs[2].set_title("Slope")
+        axs[3].plot(points, deflection(points), color='r')
```
Comment (moorepants) [inline sympy/physics/continuum_mechanics/beam.py]: I recommend cycling through matplotlib's standard color cycler instead of explicitly naming the colors.

File: sympy/physics/continuum_mechanics/beam.py [hunk 12]
```diff
@@ -1116,6 +1119,297 @@ def max_deflection(self):
...
+            length = self.length
+
+        # As we are using matplotlib directly in this method, we need to change SymPy methods
+        # to numpy functions.
+        shear = lambdify(variable, self.shear_force().subs(subs).rewrite(Piecewise), 'numpy')
+        moment = lambdify(variable, self.bending_moment().subs(subs).rewrite(Piecewise), 'numpy')
+        slope = lambdify(variable, self.slope().subs(subs).rewrite(Piecewise), 'numpy')
+        deflection = lambdify(variable, self.deflection().subs(subs).rewrite(Piecewise), 'numpy')
+
+        points = linspace(0, float(length))
```
Comment (moorepants) [inline sympy/physics/continuum_mechanics/beam.py]: I suggest setting the option `num=` to be a function of the length. You may want to guarantee some standard number of points per unit length, so plots always look good.

Comment (jashan498) [reply sympy/physics/continuum_mechanics/beam.py]: Added `num=5*length`

**Comment by jashan498** [comment]
@moorepants @parsoyaarihant is there anything else that needs to be changed?

**Comment by moorepants** [comment]
Did you address my comments on the signs of the plots so that they look like the textbook snippet I shared?

File: sympy/physics/continuum_mechanics/beam.py (hunk 6 above)
Comment (moorepants) [reply sympy/physics/continuum_mechanics/beam.py]: Ok that's fine.

**Comment by jashan498** [comment]
Yeah, see [this](https://github.com/sympy/sympy/pull/14967#issuecomment-409309807) comment.

**Comment by moorepants** [comment]
Ok great. Let's merge this.

**Review by moorepants — APPROVED** [review]
